upload.py

```diff
@@ -33,55 +35,80 @@ def hrsize(size):
             return '%.0f%sB' % (size, suffix)
     return '%.0f%sB' % (size, 'Y')
 
+
+class HashingFile(file):
+    def __init__(self, file_path):
+        super(HashingFile, self).__init__(file_path, "w+b")
+        self.sha1 = hashlib.sha1()
+
+    def write(self, data):
+        self.sha1.update(data)
+        return file.write(self, data)
+
+    def get_hash(self):
+        return self.sha1.hexdigest()
+
+
+def getHashingFieldStorage(upload_dir):
+    class HashingFieldStorage(cgi.FieldStorage):
+        def make_file(self, binary=None):
+            self.open_file = HashingFile(os.path.join(upload_dir, self.filename))
+            return self.open_file
+
+        def get_hash(self):
+            return self.open_file.get_hash()
+
+    return HashingFieldStorage
+
+
 class Upload(webapp2.RequestHandler):
 
     def get(self):
         self.response.write('Simple uploader\n')
 
-    def post(self, filename='upload.dat'):
-        return self.put(filename)
+    def post(self):
+        return self.put()
+
+    def put(self):
+        before = resource.getrusage(resource.RUSAGE_SELF)
+        before_io = psutil.disk_io_counters()
 
-    def put(self, filename='upload.dat'):
         upload_source = '%s (%s)' % (self.request.user_agent, self.request.client_addr)
         log.debug('incoming upload from ' + upload_source)
-        if self.request.content_type == 'multipart/form-data':
-            filestream = None
-            # use cgi lib to parse multipart data without loading all into memory; use tempfile instead
-            # FIXME avoid using tempfile; processs incoming stream on the fly
+        log.debug('type: ' + self.request.content_type)
+
+        if (self.request.content_type == 'multipart/form-data'):
             fs_environ = self.request.environ.copy()
             fs_environ.setdefault('CONTENT_LENGTH', '0')
             fs_environ['QUERY_STRING'] = ''
-            form = cgi.FieldStorage(fp=self.request.body_file, environ=fs_environ, keep_blank_values=True)
-            for fieldname in form:
-                field = form[fieldname]
-                if fieldname == 'file':
-                    filestream = field.file
-                    filename = field.filename
-                elif fieldname == 'metadata':
-                    try:
-                        metadata = json.loads(field.value)
-                        log.info('metadata: %s' % str(metadata))
-                    except ValueError:
-                        self.abort(400, 'non-JSON value in "metadata" parameter')
-            if filestream is None:
-                self.abort(400, 'multipart/form-data must contain a "file" field')
+
+            # Any incoming file(s) are hashed and written to disk on construction of the HashingFieldStorage class
+            form = getHashingFieldStorage(self.app.path)(fp=self.request.body_file, environ=fs_environ, keep_blank_values=True)
```

Comment:
Explain this line? I don't understand the second call (beginning with `fp`). Since it's a class object, what effect does calling it do? Can there be a default method associated with a class in python?

Also, I understand that you're creating a `HashingFieldStorage` which inherits from `cgi.FieldStorage`.
Did we only have to overwrite the `make_file()` function to fix everything?

> The class is subclassable, mostly for the purpose of overriding
> the make_file() method, which is called internally to come up with
> a file open for reading and writing.  This makes it possible to
> override the default choice of storing all files in a temporary
> directory and unlinking them as soon as they have been opened.

If so, doc?